src/slurm_plugin/clustermgtd.py

```diff
@@ -964,6 +973,73 @@ def _is_node_in_replacement_valid(self, node, check_node_is_valid):
             return not time_is_expired if check_node_is_valid else time_is_expired
         return False
 
+    def _handle_ice_nodes(self, active_nodes, unhealthy_dynamic_nodes):
+        ice_nodes = [node for node in unhealthy_dynamic_nodes if node.is_ice()]
+        self.update_insufficient_capacity_compute_resources()
+        self.reset_timeout_expire_compute_resources(ice_nodes)
+        self._set_ice_crs_to_down(active_nodes)
+
+    def update_insufficient_capacity_compute_resources(self, ice_nodes):
+        """Add compute resource to insufficient_capacity_compute_resources if node is ICE node."""
+        for node in ice_nodes:
+            compute_resource = node.get_compute_resource_name()
+            if compute_resource not in self._insufficient_capacity_compute_resources:
+                self._insufficient_capacity_compute_resources[compute_resource] = {
+                    "ice_time": self._current_time,
+                    "error_code": node.error_code,
+                }
+
+    def reset_timeout_expire_compute_resources(self, ice_nodes):
+        """Reset compute resources which insufficient_capacity_timeout expire."""
+        # Find insufficient_capacity_timeout compute resources
+        if not self._insufficient_capacity_compute_resources:
+            return
+        log.info(
+            "Insufficient capacity compute resources: %s, compute resources will be reset after "
+            "insufficient capacity timeout(%s seconds) expire",
+            self._insufficient_capacity_compute_resources,
+            self._config.insufficient_capacity_timeout,
+        )
+        timeout_expire_cr = {}
+        for cr, attribute in self._insufficient_capacity_compute_resources.copy().items():
+            if time_is_up(attribute.get("ice_time"), self._current_time, self._config.insufficient_capacity_timeout):
+                timeout_expire_cr[cr] = self._insufficient_capacity_compute_resources.pop(cr)
+
+        # Reset nodes which insufficient capacity timeout expire
+        if timeout_expire_cr:
+            logging.info(
+                f"Reset the following compute resources because insufficient capacity timeout expire: {timeout_expire_cr}"
+            )
+            nodes_to_power_down = [node for node in ice_nodes if node.get_compute_resource_name() in timeout_expire_cr]
```

Comment:
you don't need to loop over all nodes again if you change the way you store this. See my other comment above.